engine/src/main/scala/cromwell/engine/workflow/WorkflowDockerLookupActor.scala

```diff
@@ -0,0 +1,322 @@
+  * 4) Failure to look up a docker hash.
+  * 5) Timeout from DockerHashActor.
+  *
+  * Behavior:
+  * 1-2) Return a terminal lookup failure for all pending requests, transition to a permanently Failed state in which any
+  *      future requests will immediately return lookup failure.  The JobPreparation actor should fail in response to this
+  *      lookup termination message, which in turn should fail the workflow.
+  * 3-4) Return lookup failure for the current request and all pending requests for the same tag.  Any future requests
+  *      for this tag will be attempted again.
+  * 5)   Return lookup failure for all pending requests.  Any future requests for these tags will be attempted again.
+  */
+
+class WorkflowDockerLookupActor private[workflow](workflowId: WorkflowId, val dockerHashingActor: ActorRef, startMode: StartMode, databaseInterface: SqlDatabase)
+  extends LoggingFSM[WorkflowDockerLookupActorState, WorkflowDockerLookupActorData] with DockerClientHelper with SingletonServicesStore {
+
+  implicit val ec = context.system.dispatchers.lookup(Dispatcher.EngineDispatcher)
+
+  // Amount of time after which the docker request should be considered lost and sent again.
+  override protected def backpressureTimeout: FiniteDuration = 10 seconds
+  // Amount of time to wait when we get a Backpressure response before sending the request again.
+  override protected def backpressureRandomizerFactor: Double = 0.5D
+
+  context.become(dockerReceive orElse receive)
+
+  {
+    val restart = startMode == RestartExistingWorkflow
+    val initialState = if (restart) AwaitingFirstRequestOnRestart else Running
+
+    startWith(initialState, WorkflowDockerLookupActorData.empty)
+  }
+
+  // `AwaitingFirstRequestOnRestart` is only used in restart scenarios.  This state waits until there's at least one hash
+  // request before trying to load the docker hash mappings.  This is so we'll have at least one `JobPreparationActor`
+  // reference available to message with a terminal failure in case the reading or parsing of these mappings fails.
+  when(AwaitingFirstRequestOnRestart) {
+    case Event(request: DockerHashRequest, _) =>
+      loadDockerHashStoreEntries()
+      self forward request
+      goto(LoadingCache)
+  }
+
+  // Waiting for a response from the database with the hash mapping for this workflow.
+  when(LoadingCache) {
+    case Event(request: DockerHashRequest, data) =>
+      stay using data.addHashRequest(request, sender())
+    case Event(DockerHashStoreLoadingSuccess(dockerHashEntries), data) =>
+      loadCacheAndHandleHashRequests(dockerHashEntries, data)
+  }
+
+  // This is the normal operational mode.
+  when(Running) {
+    // This tag has already been looked up and its hash is in the mappings cache.
+    case Event(request: DockerHashRequest, data) if data.mappings.contains(request.dockerImageID) =>
+      sender ! DockerHashSuccessResponse(data.mappings(request.dockerImageID), request)
```

Comment:
Yeah but then two distinct cases (mappings in cache vs mappings not in cache) would be smushed into one condition, which I think other reviewers might not like.  😄